### PR TITLE
feat: reload strategy

### PR DESCRIPTION
**Pull request by mzz2017**
```bash
systemctl reload mmp-go
```

Use USR1 signal to inform.

### CODE REVIEW

File: systemd/mmp-go.service [hunk 1]
```diff
@@ -13,6 +13,7 @@ CapabilityBoundingSet=CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_RAW
 AmbientCapabilities=CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_RAW
```
Comment (DuckSoft) [inline systemd/mmp-go.service:13]: CAP_NET_RAW CAP_NET_ADMIN 是不是多余……

File: signal_windows.go [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package main
+
+// not support windows
+func signalHandler() {}
```
Comment (DuckSoft) [inline signal_windows.go:4]: 建议打个警告之类的出来

File: systemd/mmp-go.service (hunk 1 above)
Comment (mzz2017) [reply systemd/mmp-go.service:13]: 是的，但应该不在本pr的范围。。？

File: signal_windows.go (hunk 2 above)
Comment (mzz2017) [reply signal_windows.go:4]: 怎么说？

File: reload.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	for port := range mPortDispatcher {
+		if _, ok := newConfPortSet[port]; !ok {
+			t := mPortDispatcher[port]
+			delete(mPortDispatcher, port)
+			for j := range protocols {
+				_ = (*t)[j].Close()
+			}
+		}
+	}
+	log.Println("configure reloaded")
```
Comment (DuckSoft) [inline reload.go:54]:
```suggestion
	log.Println("configuration reloaded")
```

File: reload.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+func ReloadConfig() {
+	log.Println("configure reloading")
+	mMutex.Lock()
+	defer mMutex.Unlock()
+
+	// rebuild config
+	confPath := config.GetConfig().ConfPath
+	newConf, err := config.BuildConfig(confPath)
+	if err != nil {
+		log.Printf("failed to reload: %v", err)
```
Comment (DuckSoft) [inline reload.go:22]:
```suggestion
		log.Printf("failed to reload configuration: %v", err)
```

also consider abort?

**Review by DuckSoft — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 5]
```diff
@@ -5,7 +5,7 @@
 *.so
 *.dylib
 /shadomplexer-go
-/mmp-go
+/mmp-go*
```
Comment (DuckSoft) [inline .gitignore:8]: you may also want to add *.exe

File: main.go [hunk 6]
```diff
@@ -10,27 +10,47 @@ import (
...
-			if err != nil {
-				log.Fatalln(err)
-			}
-			wg.Done()
-		}(&conf.Groups[i])
+		go listen(&conf.Groups[i])
 	}
+	mMutex.Unlock()
 	wg.Wait()
+	log.Println("quit")
```
Comment (DuckSoft) [inline main.go:27]: is this a debug print? if so, remove this. or you should print prettier (maybe)

File: systemd/mmp-go.service (hunk 1 above)
Comment (mzz2017) [reply systemd/mmp-go.service:13]: 这条PR里一起改了算了

File: main.go [hunk 7]
```diff
@@ -10,27 +10,47 @@ import (
...
 }
 
-func listen(group *config.Group, protocols []string) error {
+func listenWithProtocols(group *config.Group, protocols []string) error {
 	ch := make(chan error, len(protocols))
-	for _, protocol := range protocols {
+	for i, protocol := range protocols {
 		d, _ := dispatcher.New(protocol, group)
+		t := mPortDispatcher[group.Port]
+		(*t)[i] = d
```
Comment (DuckSoft) [inline main.go:53]: ummm... inline these?

File: reload.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/Qv2ray/mmp-go/dispatcher"
+	"log"
+	"sync"
+)
+
+var mMutex sync.Mutex
+var mPortDispatcher = make(map[int]*[len(protocols)]dispatcher.Dispatcher)
+
+func ReloadConfig() {
+	log.Println("configure reloading")
```
Comment (DuckSoft) [inline reload.go:14]:
```suggestion
	log.Println("reloading configuration")
```

File: .gitignore (hunk 5 above)
Comment (mzz2017) [reply .gitignore:8]: https://github.com/Qv2ray/mmp-go/blob/2e803babf22f1824e7f0d6a71a3010bad28d1b5b/.gitignore#L2

File: signal_windows.go (hunk 2 above)
Comment (DuckSoft) [reply signal_windows.go:4]: like: `Signal-triggered configuration reloading is not supported on Windows`

Comment (mzz2017) [reply signal_windows.go:4]: You mean as an annotate or an output?

Comment (DuckSoft) [reply signal_windows.go:4]: Print out a warning.

Comment (mzz2017) [reply signal_windows.go:4]: We cannot capture signal on windows, so you mean that we print it every start?

**Pull request merged**